libclang parses like a C compiler

I looks like libclang incorporates a "C" compiler when parsing the sources.
This is a problem since the generated `mockup.cc` file will be compiled as C++; headers may parse differently dependent on the mode.

For example, the `bool` type from <stdbool.h> is #defined to int when parsed by libclang, but is a native type _Bool in C99 and C++ mode.

```
[build] /build/some_test/mockup.cc:21:5: error: functions that differ only in their return type cannot be overloaded
[build] int func_x(uint8_t arg){
[build] ~~~ ^
[build] /project/include/some_include.h:65:6: note: previous declaration is here
[build] bool func_x(uint8_t arg);
[build] ~~~~ ^
```
We need to control the compiler that libclang incorporates to be the same C++ compiler that will compile the code.

--- a/hammocking/hammocking.py
+++ b/hammocking/hammocking.py
@@ -114,6 +114,9 @@ class MockupWriter:
         template = Path(Path(template_filename).stem)
         return template.stem + (self.suffix if self.suffix else "") + template.suffix
 
+    def default_language_mode(self) -> str:
+        return 'c++' if self.mockup_style in ["gmock"] else 'c'
+
 
 class Hammock:
     def __init__(self, symbols: List[str], cmd_args: List[str] = [], mockup_style="gmock", suffix=None):
@@ -134,6 +137,9 @@ class Hammock:
             "args": self.cmd_args,
             "options": TranslationUnit.PARSE_SKIP_FUNCTION_BODIES | TranslationUnit.PARSE_INCOMPLETE,
         }
+        # Determine language mode, if not explicitly given
+        if not any(arg.startswith('-x') for arg in parseOpts["args"]):
+            parseOpts["args"].append('-x' + self.writer.default_language_mode())
 
         if issubclass(type(input), Path):
             # Read a path

--- a/tests/data/mini_c++_test/some_header.h
+++ b/tests/data/mini_c++_test/some_header.h
@@ -1,0 +1,3 @@
+#include "stdbool.h"
+
+extern bool bool_status(void);

--- a/tests/data/mini_c++_test/stdbool.h
+++ b/tests/data/mini_c++_test/stdbool.h
@@ -1,0 +1,54 @@
+/* Copyright (C) 1998-2019 Free Software Foundation, Inc.
+
+This file is part of GCC.
+
+GCC is free software; you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation; either version 3, or (at your option)
+any later version.
+
+GCC is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+Under Section 7 of GPL version 3, you are granted additional
+permissions described in the GCC Runtime Library Exception, version
+3.1, as published by the Free Software Foundation.
+
+You should have received a copy of the GNU General Public License and
+a copy of the GCC Runtime Library Exception along with this program;
+see the files COPYING3 and COPYING.RUNTIME respectively.  If not, see
+<http://www.gnu.org/licenses/>.  */
+
+/*
+ * ISO C Standard:  7.16  Boolean type and values  <stdbool.h>
+ */
+
+#ifndef _STDBOOL_H
+#define _STDBOOL_H
+
+#ifndef __cplusplus
+
+#define bool    _Bool
+#define true    1
+#define false   0
+
+#else /* __cplusplus */
+
+/* Supporting _Bool in C++ is a GCC extension.  */
+#define _Bool   bool
+
+#if __cplusplus < 201103L
+/* Defining these macros in C++98 is a GCC extension.  */
+#define bool    bool
+#define false   false
+#define true    true
+#endif
+
+#endif /* __cplusplus */
+
+/* Signal that all the definitions are present.  */
+#define __bool_true_false_are_defined   1
+
+#endif  /* stdbool.h */

--- a/tests/data/mini_c++_test/use_bool.c
+++ b/tests/data/mini_c++_test/use_bool.c
@@ -1,0 +1,5 @@
+#include "some_header.h"
+
+int func_a(void) {
+    return bool_status() ? 1: 0;
+}

--- a/tests/hammocking_test.py
+++ b/tests/hammocking_test.py
@@ -187,6 +187,12 @@ float y;
         assert writer.get_mockup('mockup.h') == open("tests/data/gmock_test/test_mini_c_gmock/mockup.h").read()
         assert writer.get_mockup('mockup.cc') == open("tests/data/gmock_test/test_mini_c_gmock/mockup.cc").read()
 
+    def test_languagemode(self):
+        writer = MockupWriter()
+        assert writer.default_language_mode() == 'c++'
+        writer.set_mockup_style('plain_c')
+        assert writer.default_language_mode() == 'c'
+
 
 class TestNmWrapper(unittest.TestCase):
     regex = NmWrapper.regex
@@ -287,6 +293,22 @@ extern void ignore_me();
         assert mock.writer.variables[0].get_definition() == "int b", "Variable shall be created in the mockup"
         assert len(mock.writer.functions) == 1, "Mockup shall have a function"
         assert mock.writer.functions[0].get_signature() == "void foo()", "Function shall be created in the mockup"
+
+    def test_langmode_auto(self):
+        """Read as c++ compiler determined from output style"""
+        mock = Hammock(["bool_status"])
+        mock.parse(Path("tests/data/mini_c++_test/use_bool.c"))
+        assert mock.done, "Should be done now"
+        assert len(mock.writer.functions) == 1, "Mockup shall have a function"
+        assert mock.writer.functions[0].get_signature() == "bool bool_status()", "Function shall be created with bool type"
+
+    def test_langmode_override(self):
+        """Read as c compiler"""
+        mock = Hammock(["bool_status"], ["-xc"])
+        mock.parse(Path("tests/data/mini_c++_test/use_bool.c"))
+        assert mock.done, "Should be done now"
+        assert len(mock.writer.functions) == 1, "Mockup shall have a function"
+        assert mock.writer.functions[0].get_signature() == "_Bool bool_status()", "Function shall be created with C99 bool type"
 
 
 if __name__ == "__main__":

